designs/0034-jacobian.md

```diff
@@ -0,0 +1,320 @@
+
+Having a new block for user-defined constraints allows users and developers to remove some of the veil of encapsulation and write repeatable code for future models.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The following section is a draft of the docs for the new constraint and `jacobian` keyword.
+
+## Constraint Block
+
+### Overview
+
+The constraints block was included in Stan 2.XX, introduced to provide users with the ability to define custom constraints on parameters, transformed parameters, and data. This block is situated immediately after the `functions` block and before the `data` block in a Stan program. The primary purpose of the `constraints` block is to allow users to define custom constraints that can be used in the same way as built-in constraints.
+
+### Syntax
+
+A constraints block is defined by the keyword `constraints`. Inside of the `constraints` block each constraint is its own block with a given `constraint_name` and the functions needed to define the constraint. Inside the `constraints` block, users can define one or more constraint types. Each constraint type is encapsulated within its own set of curly braces and is identified by a unique name.
+
+Each user-defined constraint within the constraint block must specify a set of functions that define the transformation between unconstrained and constrained spaces, validate data, and increment the log of the absolute value of the determinant of the Jacobian. The general structure of a user-defined constraint is as follows:
+
+```stan
+constraints {
+ constraint_name {
+   // Transform from unconstrained to constrained space
+   ReturnType constrain(TransformType, OtherArgs...) {
+     // ... transformation logic ...
+     // Increment Jacobian
+     jacobian += ...;
+     return ...;
+   }
+
+
+   // Transform from constrained to unconstrained space
+   type unconstrain(TransformType, OtherArgs...) {
+     // ... inverse transformation logic ...
+     return ...;
+   }
+
+
+   // Validate data
+   int validate(TransformType, OtherArgs...) {
+     // ... validation logic ...
+     return ...; // typically returns a boolean value
+   }
+ }
+}
+```
+
+Each constraint type must define a set of functions that govern the transformation and validation of variables subject to the constraint. The first argument for all functions defined in a constraint must be the argument of interest to be transformed or validated. These functions include:
+
```

Comment:
This is confusingly mixing types (`TransformType`) and arguments (`OtherArguments`).  How about putting it more in Stan syntax, as in:

```
C constrain(U x, T1 a1, ..., TN aN);
```

and say that `U` is the unconstrained type, `C` is the constrained type, and `T1...TN` are types of extra arguments.

Does the constrained type U have to be a vector?

Reply:
This might sound silly, but if I make the signature like 

```
C constrain(U x, T1 a1, ..., TN aN);
```

Users will think the signatures arguments have to be `x`, `a1`, `aN` etc? 

I added a section to this that defines what `ReturnType`, `UnconstrainType`, and `OtherArgs` means